hack/setup-envtest-with-kind.sh

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+
+# Copyright 2021 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+os="unknown"
+if [[ "${OSTYPE}" == "linux"* ]]; then
+  os="linux"
+elif [[ "${OSTYPE}" == "darwin"* ]]; then
+  os="darwin"
+fi
+
+if [[ "$os" == "unknown" ]]; then
+  echo "OS '$OSTYPE' not supported. Aborting." >&2
+  exit 1
+fi
+
+echo "Execute the following commands to prepare your environment to run the tests with kind:"
+echo ""
+echo "kind create cluster"
+echo "export USE_EXISTING_CLUSTER=true"
+
+if [[ "${os}" == "linux" ]]; then
+  echo "gateway_ip=\$(docker inspect kind-control-plane | jq '.[0].NetworkSettings.Networks.kind.Gateway' -r)"
+  echo "docker exec kind-control-plane bash -c \"echo \\\"\${gateway_ip} linux.localhost\\\" >> /etc/hosts\""
+  echo "export CAPI_WEBHOOK_HOSTNAME=\"linux.localhost\""
+elif [[ "${os}" == "darwin" ]]; then
+  echo "export CAPI_WEBHOOK_HOSTNAME=\"docker.for.mac.localhost\""
+fi
```

Comment:
Can we make these steps a bit more automated? We could offer some `Makefile` commands to achieve something like this by using a well defined `kind` cluster name.

For example, let's assume we can use the name `cluster-api-envtest`.

### make envtest-start
Creates the kind cluster with the name above, we can check first if the cluster is running before creating a new one.

### make clean-envtest
Removes the kind cluster above.

If we do the above, assuming we can easily check whenever a cluster is running, we can modify `make test` to auto-export the environment variables and log out that we're using the already-running kind cluster.

Reply:
Yup, will do. I have to take a look how much scripting it takes to make it robust. (because existence of kind cluster != that it's also started)

Reply:
Sounds good, it can be definitely worth it for a nicer UX

Reply:
There's a problem with using `make test` (aka `go test ./...`). 

It runs all our test-suites / packages in parallel. This works fine when every package uses its own test env. When all are using the same kind cluster the tests fail because (at least) the cluster-wide webhook configurations overwrite each other (especially the URL of the webhook). Thus the tests fail because the webhooks are not always reachable.

That's not a problem in my use case which is to run the tests of only one package at the same time. So what I can do is create a `envtest-start` and `envtest-clean` target. For test execution I think we have the following options:
* print out the export commands and an example `go test ./controllers/...` command at the end of `envtest-start`
* Add a new make target `envtest-package`. This target would automatically export the environment variables as you've suggested and it would also take a package path via environmnet variable as parameter, e.g. `PACKAGE=./controllers make envtest-package`

Reply:
Could we add some checks that if the webhooks are already registered, we don't overwrite them again?

Reply:
We can. But if the first package finishes then other packages  will fail because the webhook server has been shutdown already.

The only clean solution I can think of is to make sure our webhooks only apply to the current package by:
* using random or namespaced names for our webhook resources
* using an object selector in the webhook

Reply:
The selector in the webhook might work only for mutating or validation, conversion webhook are registered globally so that wouldn't work.

One other thing to consider doing is having a go main file that registers webhooks separately for testenv

Reply:
So we would have a separate binary we would start which runs the webhooks, then we run the regular tests in parallel (without registering the webhooks again) and after the tests are done we stop the separate binary?

To be honest, I'm not sure if it's worth the effort (i.e. if enough folks would use it). My use case is only:
* startup kind
* run a specific test or maybe a whole package via IDE repeatedly

I'm not sure we have to support running all the tests in the repo against the same kind cluster concurrently.

Reply:
👍